onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -0,0 +1,314 @@
+// Licensed under the MIT License.
+
+#include "coreml_execution_provider.h"
+
+#include "core/framework/allocatormgr.h"
+#include "core/framework/compute_capability.h"
+#include "core/graph/graph_viewer.h"
+#include "core/session/onnxruntime_cxx_api.h"
+
+#include "model/model.h"
+#include "model/host_utils.h"
+#include "builders/helper.h"
+#include "builders/model_builder.h"
+
+namespace onnxruntime {
+
+constexpr const char* COREML = "CoreML";
+
+CoreMLExecutionProvider::CoreMLExecutionProvider()
+    : IExecutionProvider{onnxruntime::kCoreMLExecutionProvider} {
+  AllocatorCreationInfo device_info(
+      [](int) {
+        return onnxruntime::make_unique<CPUAllocator>(OrtMemoryInfo(COREML, OrtAllocatorType::OrtDeviceAllocator));
+      });
+
+  InsertAllocator(CreateAllocator(device_info));
+
+  AllocatorCreationInfo cpu_memory_info(
+      [](int) {
+        return onnxruntime::make_unique<CPUAllocator>(
+            OrtMemoryInfo(COREML, OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), 0, OrtMemTypeCPUOutput));
+      });
+
+  InsertAllocator(CreateAllocator(cpu_memory_info));
+}
+
+CoreMLExecutionProvider::~CoreMLExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+CoreMLExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_view,
+                                       const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+
+  // We do not run CoreML EP on subgraph, instead we cover this in the control flow nodes
+  if (graph_view.IsSubgraph()) {
+    return result;
+  }
+
+  std::unordered_set<std::string> all_node_inputs;
+  for (const auto& node : graph_view.Nodes()) {
+    for (auto* input : node.InputDefs()) {
+      all_node_inputs.insert(input->Name());
+    }
+  }
+
+  const auto supported_nodes_vector = coreml::GetSupportedNodes(graph_view);
+
+  // Find inputs, initializers and outputs for each supported subgraph
+  size_t num_of_supported_nodes = 0;
+  const std::vector<NodeIndex>& node_index = graph_view.GetNodesInTopologicalOrder();
+  const auto& graph_outputs = graph_view.GetOutputs();
+  for (const auto& group : supported_nodes_vector) {
+    if (group.empty())
+      continue;
+
+    num_of_supported_nodes += group.size();
+    LOGS_DEFAULT(VERBOSE) << "CoreMLExecutionProvider::GetCapability, current supported node group size: "
+                          << group.size();
+
+    std::unordered_set<size_t> node_set;
+    node_set.reserve(group.size());
+    for (const auto& index : group) {
+      node_set.insert(node_index[index]);
+    }
+
+    std::unique_ptr<IndexedSubGraph> sub_graph = onnxruntime::make_unique<IndexedSubGraph>();
+    // Find inputs and outputs of the subgraph
+    std::unordered_map<const NodeArg*, int> fused_inputs, fused_outputs, fused_outputs_to_add;
+    std::unordered_set<const NodeArg*> erased;
+    int input_order = 0;
```

Comment:
Please consider using the approach in the internal testing EP for determining the inputs/outputs. I believe it is the most concise way to do it, and is hopefully easy to understand.

[https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/providers/internal_testing/internal_testing_execution_provider.cc#L88](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/providers/internal_testing/internal_testing_execution_provider.cc#L88)